Config: coplien's form + multi CGI extensions per location

edouardproust opened the pull request:
- Added coplien's form to classes of the `config` module: `Config`, `ServerBlock`, `LocationBlock`
- Removed attribute `ServerBock::server_name` and the logic around it (virtual host)
- Updated cgi attributes in `LocationBlock` to allow several extensions per location
- Added global headers `inc/constants.hpp` and `inc/typedefs.hpp`
- Updated somes checks in `config` classes `validate()` (eg. now `root` is not mandatory in `ServerBlock` and will be empty if not set)

edouardproust committed:
implemented coplien's form for config module

edouardproust committed:
did coplien form + removed virtual host logic

edouardproust committed:
better config module

Skoteini-42 commented:
Nice idea about the global headers inc/constants.hpp and inc/typedefs.hpp, all good for me! Don't forget to delete the extra functions I put in main for the parse tester after you merge this and good to go :)

edouardproust committed:
merge with html_base

edouardproust committed:
merged with http_base

edouardproust commented:
Update:
- Moved some variables of module `http` into `constants.hpp`
- Added a rule `dev` in `Makefile` for testing purpose during development.
Files are added in `Makefile` under `DEV_SRC_FILES` and should be named starting with `dev.*` ( I add `dev.*` to `..gitignore`) and functions be located in namespace `dev`. For more info, check `inc/http/dev.http.hpp` and `src/http/dev.http.cpp`

To run the dev version:
```
make dev
./webserv_dev tests/config/basic.config
```
In a .cpp file, we can now test things using `DEVMODE` variable, like I did in `main.cpp`:
```
if (DEVMODE) {
  cfg.print();
  dev::runParserTests();
}

edouardproust committed:
cleaned headers

edouardproust committed:
http and router logic base

edouardproust committed:
update LocationBlock with Request::isSupportedMethod()

edouardproust commented:
Ava i put what we talked about. I added a `Router` class as a base so you can see a bit the logic of the params the constructor take etc. It serve also as a base for me to start working on the Router.

I also added what you validated: the `Request::_supportedMethods` list and `Request::isSupportedMethod()`

I also changed the tests made in `main.cpp` like you saw previously. I removed the `try/catch`, you will decide what is best for handling the case of a badly formated raw request.

(of course the `RequestParser` will be executed inside the server, not in `main`, the block `if (DEVMODE)` in `main` is for testing purpose right now)

If you want and if you're quick to finish the `http` module, you'll be able to switch to `router` and finish it so I can switch to `CGI` myself
